File: backend/app/models/quiz.py
from pydantic import BaseModel
from typing import List, Optional, Dict

class Question(BaseModel):
    id: str
    questionText: str
    type: str
    options: List[str]
    correctAnswerIndex: Optional[int] = None
    correctAnswerIndices: Optional[List[int]] = None
    dragItems: Optional[List[str]] = None
    dropTargets: Optional[List[str]] = None
    correctMatches: Optional[Dict[str, str]] = None

class Quiz(BaseModel):
    id: Optional[str] = None
    title: str
    description: str
    language: str
    category: str
    coverImagePath: Optional[str] = None
    creatorId: str
    originalOwner: Optional[str] = None  # For quizzes added from other users
    sharedMode: Optional[str] = None  # Mode used when quiz was shared (share, self_paced, timed_individual, live_multiplayer)
    questions: List[Question]
    createdAt: Optional[str] = None  # string instead of datetime


class QuizResponse(BaseModel):
    id: str
    message: str

class QuizLibraryItem(BaseModel):
    id: str
    title: str
    description: str
    coverImagePath: Optional[str] = None
    createdAt: Optional[str] = None
    questionCount: int
    language: str
    category: str
    originalOwner: Optional[str] = None
    originalOwnerUsername: Optional[str] = None  # Fetched from Firestore
    sharedMode: Optional[str] = None  # Mode used when quiz was shared


class QuizLibraryResponse(BaseModel):
    success: bool
    data: List[QuizLibraryItem]
    count: int

File: quiz_app/lib/CreateSection/models/quiz.dart
import 'package:quiz_app/CreateSection/models/question.dart';

class Quiz {
  String? id;
  String title;
  String description;
  String language;
  String category;
  String? coverImagePath;
  String creatorId;
  List<Question> questions;
  DateTime createdAt;

  Quiz({
    this.id,
    required this.title,
    required this.description,
    required this.language,
    required this.category,
    this.coverImagePath,
    required this.creatorId,
    List<Question>? questions,
    DateTime? createdAt,
  }) : questions = questions ?? [],
       createdAt = createdAt ?? DateTime.now();

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'language': language,
      'category': category,
      'coverImagePath': coverImagePath,
      'creatorId': creatorId,
      'questions': questions.map((q) => q.toJson()).toList(),
      'createdAt': createdAt.toIso8601String(),
    };
  }

  factory Quiz.fromJson(Map<String, dynamic> json) {
    return Quiz(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      language: json['language'],
      category: json['category'],
      coverImagePath: json['coverImagePath'],
      creatorId: json['creatorId'] ?? '',
      questions:
          (json['questions'] as List).map((q) => Question.fromJson(q)).toList(),
      createdAt: DateTime.parse(json['createdAt']),
    );
  }
}
